### PR TITLE
[wasm][debugger] Symbolicate function names on debug callstack

### PR DESCRIPTION
**Pull request by thaystg**
Running the app with Just My Code disabled.
Without the PR a browserwasm app:
image

With the PR a browserwasm app:
image

### CODE REVIEW

File: src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs [hunk 1]
```diff
@@ -477,6 +477,7 @@ public DebugStore Store
         }
 
         public bool DisableSymbolicate { get; internal set; }
+        public string[] Symbols { get; internal set; }
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:480]: `Symbols` -> `WasmFunctionIds`?

File: src/mono/wasm/runtime/logging.ts [hunk 2]
```diff
@@ -148,4 +148,9 @@ export function parseSymbolMapFile(text: string) {
     });
 
     mono_log_debug(`Loaded ${wasm_func_map.size} symbols`);
+}
+
```
Comment (radical) [inline src/mono/wasm/runtime/logging.ts:152]:
```suggestion
```

File: src/mono/wasm/runtime/logging.ts [hunk 3]
```diff
@@ -148,4 +148,9 @@ export function parseSymbolMapFile(text: string) {
...
+
+export function mono_wasm_get_symbols() {
```
Comment (radical) [inline src/mono/wasm/runtime/logging.ts:154]: Maybe: `mono_wasm_get_func_id_to_name_mappings`?

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 4]
```diff
@@ -1157,6 +1153,30 @@ protected virtual async Task<bool> SendCallStack(SessionId sessionId, ExecutionC
             return true;
         }
 
+        private async Task<string> SymbolicateFunctionName(SessionId sessionId, ExecutionContext context, int funcId, CancellationToken token)
```
[the remark this answers is not in the document]
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:1156]: I don't think so, this is called when we are mounting the callstack, so we are already paused in a breakpoint or in an exception.

File: src/mono/wasm/runtime/logging.ts [hunk 5]
```diff
@@ -148,4 +148,8 @@ export function parseSymbolMapFile(text: string) {
     });
 
     mono_log_debug(`Loaded ${wasm_func_map.size} symbols`);
+}
+
+export function mono_wasm_get_func_id_to_name_mappings() {
+    return [...wasm_func_map.values()];
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/logging.ts:154]: Could this rather return URL of the symbol files ? Or perhaps it's we could teach DebugProxy to read the blazor.boot.json and resolve .PDBs and DLLs in the same way too ?

Comment (pavelsavara) [reply src/mono/wasm/runtime/logging.ts:154]: In Net9 maybe

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 6]
```diff
@@ -1126,6 +1126,10 @@ protected virtual async Task<bool> SendCallStack(SessionId sessionId, ExecutionC
                         function_name.StartsWith("_mono_wasm_fire_debugger_agent_message", StringComparison.Ordinal) ||
                         function_name.StartsWith("mono_wasm_fire_debugger_agent_message", StringComparison.Ordinal)))
                 {
+                    if (function_name.StartsWith("$func", StringComparison.Ordinal) && int.TryParse(function_name.Remove(0, 5), out var funcId))
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:1129]: Can we move this condition into `SymbolicateFunctionName` itself, so references to the `"$func"` string are in one place?

File: src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs [hunk 7]
```diff
@@ -473,6 +475,7 @@ public DebugStore Store
                 return store;
             }
         }
+        public string[] GetWasmFunctionIds { get; internal set; }
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:478]:
```suggestion
        public string[] WasmFunctionIds { get; internal set; }
```

**Review by ilonatommy — APPROVED** [review]

**Pull request merged**